vAttach: get vAttach working with multi-process

xobs opened the pull request:
### Description

<!-- Please include a brief description of what is being added/changed -->

This PR implements one possible approach to getting `vAttach` working with multiple processes. With this, it is now possible to attach to a given process, for example by issuing `attach 7` to attach to PID 7.

### API Stability

- [X] This PR does not require a breaking API change

### Checklist

<!-- CI takes care of a lot of things, but there are some things that have yet to be automated -->

- Documentation
  - [X] Ensured any public-facing `rustdoc` formatting looks good (via `cargo doc`)
  - [X] (if appropriate) Added feature to "Debugging Features" in README.md
- Validation
  - [ ] Included output of running `examples/armv4t` with `RUST_LOG=trace` + any relevant GDB output under the "Validation" section below
  - [ ] Included output of running `./example_no_std/check_size.sh` before/after changes under the "Validation" section below

<!-- Oh, and if you're integrating `gdbstub` in an open-source project, do consider updating the README.md's "Real World Examples" section to link back to your project! -->

### Validation

<details>
<summary>GDB output</summary>

```
(gdb) tar ext :3456
Remote debugging using :3456
[remote] start_remote_1: enter
  [remote] Sending packet: $qSupported:multiprocess+;swbreak+;hwbreak+;qRelocInsn+;fork-events+;vfork-events+;exec-events+;vContSupported+;QThreadEvents+;no-resumed+;memory-tagging+#ec
  [remote] Received Ack
  [remote] Packet received: PacketSize=1000;vContSupported+;multiprocess+;QStartNoAckMode+;qXfer:features:read+
  [remote] packet_ok: Packet qSupported (supported-packets) is supported
  [remote] Sending packet: $vMustReplyEmpty#3a
  [remote] Received Ack
  [remote] Packet received:
  [remote] Sending packet: $QStartNoAckMode#b0
  [remote] Received Ack
  [remote] Packet received: OK
  [remote] Sending packet: $!#21
  [remote] Packet received: OK
  [remote] Sending packet: $Hgp0.0#ad
  [remote] Packet received: E01
  [remote] Sending packet: $qXfer:features:read:target.xml:0,ffb#79
  [remote] Packet received: m<target version="1.0"><architecture>riscv:rv32</architecture></target>
  [remote] Sending packet: $qXfer:features:read:target.xml:46,ffb#b3
  [remote] Packet received: l
  [remote] Sending packet: $qTStatus#49
  [remote] Packet received:
  [remote] packet_ok: Packet qTStatus (trace-status) is NOT supported
  [remote] Sending packet: $?#3f
  [remote] Packet received: W00;
[remote] start_remote_1: exit
(gdb) att 2
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, process 2
[remote] Sending packet: $vAttach;2#38
[remote] Packet received: S00
[remote] packet_ok: Packet vAttach (attach) is supported
[remote] Sending packet: $qC#b4
[remote] Packet received: QCp02.02
[remote] Sending packet: $Hgp2.2#b1
[remote] Packet received: OK
[remote] Sending packet: $qXfer:features:read:target.xml:0,ffb#79
[remote] Packet received: m<target version="1.0"><architecture>riscv:rv32</architecture></target>
[remote] Sending packet: $qXfer:features:read:target.xml:46,ffb#b3
[remote] Packet received: l
[remote] wait: enter
  [remote] select_thread_for_ambiguous_stop_reply: enter
    [remote] select_thread_for_ambiguous_stop_reply: process_wide_stop = 0
    [remote] select_thread_for_ambiguous_stop_reply: first resumed thread is Thread 2.2
    [remote] select_thread_for_ambiguous_stop_reply: is this guess ambiguous? = 0
  [remote] select_thread_for_ambiguous_stop_reply: exit
[remote] wait: exit
[remote] Sending packet: $g#67
[remote] Packet received: 00000000b431030090f5ff7f00000000000000601000000038550300bcf5ff7f90f6ff7f050000000f0000007469636b74696d65722d736572766572000000000000000000000000000000004a010000f8faff7f680400200500000001000000feffffff040000000f0f0f0f0101010101000000040000000000000090000020c8310300
[remote] Sending packet: $qOffsets#4b
[remote] Packet received:
[remote] Sending packet: $qSymbol::#5b
[remote] Packet received:
[remote] packet_ok: Packet qSymbol (symbol-lookup) is NOT supported
[remote] Sending packet: $qfThreadInfo#bb
[remote] Packet received: mp02.02,p02.03
[remote] Sending packet: $qsThreadInfo#c8
[remote] Packet received: l
[New Thread 2.3]
0x000331c8 in xous::syscall::receive_message ()
(gdb) c
Continuing.
[remote] Sending packet: $vCont?#49
[remote] Packet received: vCont;c;C;s;S
[remote] packet_ok: Packet vCont (verbose-resume) is supported
[remote] Sending packet: $vCont;c:p2.-1#10
[remote] wait: enter
[remote] wait: exit
[remote] pass_ctrlc: enter
  [remote] interrupt: enter
  [remote] interrupt: exit
[remote] pass_ctrlc: exit
[remote] wait: enter
  [remote] Packet received: S02
  [remote] select_thread_for_ambiguous_stop_reply: enter
    [remote] select_thread_for_ambiguous_stop_reply: process_wide_stop = 0
    [remote] select_thread_for_ambiguous_stop_reply: first resumed thread is Thread 2.2
    [remote] select_thread_for_ambiguous_stop_reply: is this guess ambiguous? = 1
warning: multi-threaded target stopped without sending a thread-id, using first non-exited thread
  [remote] select_thread_for_ambiguous_stop_reply: exit
[remote] wait: exit
[remote] Sending packet: $g#67
[remote] Packet received: 00000000ac33030030f6ff7f00000000000000600200000038550300bcf5ff7ff0faff7f0400000029000000460020170000000041161d00000000000000000000000000020000000000000041161d00f8faff7f680400200400000001000000feffffff040000000f0f0f0f0101010101000000030000000000000090000020c0330300
[remote] Sending packet: $qfThreadInfo#bb
[remote] Packet received: mp02.02,p02.03
[remote] Sending packet: $qsThreadInfo#c8
[remote] Packet received: l

Thread 1 received signal SIGINT, Interrupt.
0x000333c0 in xous::syscall::reply_and_receive_next_legacy ()
(gdb) c
Continuing.
[remote] Sending packet: $vCont;c:p2.-1#10
Remote communication error.  Target disconnected.: No error.
(gdb) tar ext :3456
Remote debugging using :3456
[remote] start_remote_1: enter
  [remote] Sending packet: $qSupported:multiprocess+;swbreak+;hwbreak+;qRelocInsn+;fork-events+;vfork-events+;exec-events+;vContSupported+;QThreadEvents+;no-resumed+;memory-tagging+#ec
  [remote] Received Ack
  [remote] Packet received: PacketSize=1000;vContSupported+;multiprocess+;QStartNoAckMode+;qXfer:features:read+
  [remote] packet_ok: Packet qSupported (supported-packets) is supported
  [remote] Sending packet: $vMustReplyEmpty#3a
  [remote] Received Ack
  [remote] Packet received:
  [remote] Sending packet: $QStartNoAckMode#b0
  [remote] Received Ack
  [remote] Packet received: OK
  [remote] Sending packet: $!#21
  [remote] Packet received: OK
  [remote] Sending packet: $Hgp0.0#ad
  [remote] Packet received: E01
  [remote] Sending packet: $qXfer:features:read:target.xml:0,ffb#79
  [remote] Packet received: m<target version="1.0"><architecture>riscv:rv32</architecture></target>
  [remote] Sending packet: $qXfer:features:read:target.xml:46,ffb#b3
  [remote] Packet received: l
  [remote] Sending packet: $qTStatus#49
  [remote] Packet received:
  [remote] packet_ok: Packet qTStatus (trace-status) is NOT supported
  [remote] Sending packet: $?#3f
  [remote] Packet received: W00;
[remote] start_remote_1: exit
(gdb) att 2
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, process 2
[remote] Sending packet: $vAttach;2#38
[remote] Packet received: S00
[remote] packet_ok: Packet vAttach (attach) is supported
[remote] Sending packet: $qC#b4
[remote] Packet received: QCp02.02
[remote] Sending packet: $Hgp2.2#b1
[remote] Packet received: OK
[remote] Sending packet: $qXfer:features:read:target.xml:0,ffb#79
[remote] Packet received: m<target version="1.0"><architecture>riscv:rv32</architecture></target>
[remote] Sending packet: $qXfer:features:read:target.xml:46,ffb#b3
[remote] Packet received: l
[remote] wait: enter
  [remote] select_thread_for_ambiguous_stop_reply: enter
    [remote] select_thread_for_ambiguous_stop_reply: process_wide_stop = 0
    [remote] select_thread_for_ambiguous_stop_reply: first resumed thread is Thread 2.2
    [remote] select_thread_for_ambiguous_stop_reply: is this guess ambiguous? = 0
  [remote] select_thread_for_ambiguous_stop_reply: exit
[remote] wait: exit
[remote] Sending packet: $g#67
[remote] Packet received: 00000000ac33030030f6ff7f00000000000000600200000038550300bcf5ff7ff0faff7f04000000290000000b00230d00000000a72300000000000000000000000000000200000000000000a7230000f8faff7f680400200400000001000000feffffff040000000f0f0f0f0101010101000000030000000000000090000020c0330300
[remote] Sending packet: $qOffsets#4b
[remote] Packet received:
[remote] Sending packet: $qSymbol::#5b
[remote] Packet received:
[remote] packet_ok: Packet qSymbol (symbol-lookup) is NOT supported
[remote] Sending packet: $qfThreadInfo#bb
[remote] Packet received: mp02.02,p02.03
[remote] Sending packet: $qsThreadInfo#c8
[remote] Packet received: l
[New Thread 2.3]

Thread 1 stopped.
0x000333c0 in xous::syscall::reply_and_receive_next_legacy ()
(gdb) att 3
A program is being debugged already.  Kill it? (y or n) y
[remote] Sending packet: $vKill;2#6f
[remote] Packet received: OK
[remote] packet_ok: Packet vKill (kill) is supported
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, process 3
[remote] Sending packet: $vAttach;3#39
[remote] Packet received: S00
[remote] Sending packet: $qC#b4
[remote] Packet received: QCp03.02
[remote] Sending packet: $Hgp3.2#b2
[remote] Packet received: OK
[remote] Sending packet: $qXfer:features:read:target.xml:0,ffb#79
[remote] Packet received: m<target version="1.0"><architecture>riscv:rv32</architecture></target>
[remote] Sending packet: $qXfer:features:read:target.xml:46,ffb#b3
[remote] Packet received: l
[remote] wait: enter
  [remote] select_thread_for_ambiguous_stop_reply: enter
    [remote] select_thread_for_ambiguous_stop_reply: process_wide_stop = 0
    [remote] select_thread_for_ambiguous_stop_reply: first resumed thread is Thread 3.2
    [remote] select_thread_for_ambiguous_stop_reply: is this guess ambiguous? = 0
  [remote] select_thread_for_ambiguous_stop_reply: exit
[remote] wait: exit
[remote] Sending packet: $g#67
[remote] Packet received: 00000000a663010010ffff7f000000000000006000000000d07f01007831020020ffff7f40ffff7f09000000000000000000000000000000000000000000000000000000000000001cffff7f09000000010000000000000000000000000000000000000000000000000000000000000078330200000000000000000000000000d0670100
[remote] Sending packet: $qOffsets#4b
[remote] Packet received:
[remote] Sending packet: $qfThreadInfo#bb
[remote] Packet received: mp03.02,p03.03
[remote] Sending packet: $qsThreadInfo#c8
[remote] Packet received: l
[New Thread 3.3]

Thread 1 stopped.
0x000167d0 in ?? ()
(gdb) att 4
A program is being debugged already.  Kill it? (y or n) y
[remote] Sending packet: $vKill;3#70
[remote] Packet received: OK
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, process 4
[remote] Sending packet: $vAttach;4#3a
[remote] Packet received: S00
[remote] Sending packet: $qC#b4
[remote] Packet received: QCp04.02
[remote] Sending packet: $Hgp4.2#b3
[remote] Packet received: OK
[remote] Sending packet: $qXfer:features:read:target.xml:0,ffb#79
[remote] Packet received: m<target version="1.0"><architecture>riscv:rv32</architecture></target>
[remote] Sending packet: $qXfer:features:read:target.xml:46,ffb#b3
[remote] Packet received: l
[remote] wait: enter
  [remote] select_thread_for_ambiguous_stop_reply: enter
    [remote] select_thread_for_ambiguous_stop_reply: process_wide_stop = 0
    [remote] select_thread_for_ambiguous_stop_reply: first resumed thread is Thread 4.2
    [remote] select_thread_for_ambiguous_stop_reply: is this guess ambiguous? = 0
  [remote] select_thread_for_ambiguous_stop_reply: exit
[remote] wait: exit
[remote] Sending packet: $g#67
[remote] Packet received: 00000000b4aa010040fcff7f000000000000006010000000dcbc02004cfaff7f04000000180000000f000000786f75732d6e616d652d73657276657200000000000000000000000005000000652d736500f00200384c465c786f7573808080807c000000060000002d6e616d333333337c00000015d8c6f0756d65206d616e61269c0100
[remote] Sending packet: $qOffsets#4b
[remote] Packet received:
[remote] Sending packet: $qfThreadInfo#bb
[remote] Packet received: mp04.02
[remote] Sending packet: $qsThreadInfo#c8
[remote] Packet received: l

Program stopped.
0x00019c26 in xous_ticktimer::main ()
(gdb)
```

</details>

daniel5151 commented on src/stub/core_impl/extended_mode.rs:
```diff
@@ -26,6 +26,7 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
             }
             ExtendedMode::vAttach(cmd) => {
                 ops.attach(cmd.pid).handle_error()?;
+                res.write_str("S00")?;
```
I'm not sure this is the approach I'd like to take here.

The [docs](https://sourceware.org/gdb/onlinedocs/gdb/Packets.html#index-vAttach-packet) for the `vAttach` packet clearly state that you should be able to return _any_ stop reason, so hard-coding `S00` doesn't seem like the best idea from a long-term correctness POV. Different implementations may very well have different ways of "attaching" to a running process (i.e: not just `S00`), and giving them the ability to report the right info off-the-bat (vs. forcing extra round-trips) seems better.

That's not to say I don't acknowledge and believe https://github.com/daniel5151/gdbstub/issues/126#issuecomment-1466282750, but I think it's reasonable to punt those sorts of considerations to the user.

* * *

For this initial implementation, lets have this arm return a `HandlerStatus::DeferredStopReason`, and update the docs of the `attach` handler with something along the lines of "once this method resolves, GDB will immediately request a stop reason. It is up to the implementation to ensure the stop reason is accurate".

Maybe in the future, we can think about ways to add some more "guard rails" here (e.g: limiting the scope of stop reasons, tweaking the attach handler to return a stop reason directly, etc...), but that'll require some intrusive changes and experimentations.

xobs replied:
How do I get `HandlerStatus::DeferredStopReason` to request a stop reason? If I return that from that function and don't output `S00` then GDB breaks due to not having a valid PID:

```
(gdb) att 5
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, process 5
[remote] Sending packet: $vAttach;5#3b
[remote] getpkt_or_notif_sane_1: Timed out.
[remote] getpkt_or_notif_sane_1: Timed out.
[remote] getpkt_or_notif_sane_1: Timed out.
Ignoring packet error, continuing...
[remote] packet_ok: Packet vAttach (attach) is supported
[remote] Sending packet: $qC#b4
[remote] Packet received: QC
/__w/riscv-none-elf-gcc-xpack/riscv-none-elf-gcc-xpack/build/win32-x64/sources/gdb-12.1/gdb/inferior.c:304: internal-error: find_inferior_pid: Assertion `pid != 0' failed.
A problem internal to GDB has been detected,
further debugging may prove unreliable.
Quit this debugging session? (y or n) Recursive internal problem.
```

xobs replied:
If I have it return `HandlerStatus::DeferredStopReason` then it doesn't work at all, even if I add back in `res.write_str("S00")?;`

daniel5151 replied:
Ahh, indeed, I may have led you a little astray here. Rather than:

> once this method resolves, GDB will immediately request a stop reason. It is up to the implementation to ensure the stop reason is accurate

It should _really_ be:

> one this method resolves, GDB will expect to receive a stop reason corresponding to the attach event. The implementation is responsible for setting whatever internal flags it requires in order to immediately report an accurate stop reason after this method resolves.

That is to say: in much that same way that the `resume` method doesn't immediately return a stop reason - instead deferring the act of "waiting" for a stop reason to the event loop - so too should this method defer the act of reporting a stop reason to the event loop.

* * *

That said, this is a bit... weird.
I'm can't say i'm a _huge_ fan of this kind of unnecessary control flow cotorsion.

At the same time, `S00` also seems... off.

Here's what I think: can you lift the implementation of the `Base::QuestionMark` handler into a shared function (that can live in `base.rs`, as part of the `GdbStubImpl`) that also gets used in this handler? Oh, and make that function `#[inline(always)]` please.

After all - that is _effectively_ what we're doing here.

In the future, I'll think about how I want to properly support "custom" `?` and `vAttach` responses, but if we're hard-coding a "reasonable" response in `gdbstub` for the time being, I'd rather keep that response consistent between both packets.

daniel5151 commented on src/target/ext/base/multithread.rs:
```diff
@@ -110,6 +111,12 @@ pub trait MultiThreadBase: Target {
         None
     }
+
+    /// Allow the target to change the active PID from the default fake one.
```
This is an interesting approach.

On one hand, my gut reaction was "eeew, this is really the kind of thing we'd want to hide from the user. The whole point of `MultiThreadBase` is that you _shouldn't_ have to care about anything PID related"

...but given the scope of the work required to get `MultiProcessBase` up and running, and how reasonably unintrusive this method is...

I think I'm okay with this. With some tweaks:

I'd like to reframe this method as a "power user" method - i.e: tweak the docs to make it clear that unless you "know what you're doing", this isn't something you should be tweaking. Namely:

- Reword the above-the-fold docs as `Override the reported PID when running in multithread mode (default: 1)`
- Beef up the "below the fold" docs with info as to _why_ this might be useful (i.e: support extended mode attach / switch between processes when using `ExtendedMode::attach`)

daniel5151 reviewed (changes requested):
This seems like a reasonable approach.

I suspect that at some nebulous in the future, I'll sit down and think about different ways to "tighten this up", but that's something for me to mull over and think about.

* * *

Unless you've got any reservations / concerns about this approach, go on ahead and start chewing through the checklist.

I'm particularly interested to see what kind of binary size impact this change will have.

xobs commented:
I haven't been able to run the example tests in order to check binary size due to #131

xobs commented:
vAttach `cargo bloat` before:

```
    Analyzing target/release/gdbstub-nostd

File  .text    Size             Crate Name
2.4%  72.0% 10.4KiB         [Unknown] main
0.2%   4.9%    724B           gdbstub gdbstub::stub::state_machine::GdbStubStateMachineInner<gdbstub::stub::state_machine::stat...
0.1%   2.0%    301B           gdbstub gdbstub::protocol::commands::breakpoint::BasicBreakpoint::from_slice
0.1%   1.9%    274B           gdbstub gdbstub::stub::core_impl::resume::<impl gdbstub::stub::core_impl::GdbStubImpl<T,C>>::writ...
0.1%   1.8%    269B           gdbstub gdbstub::protocol::common::hex::decode_hex_buf
0.1%   1.8%    259B           gdbstub <gdbstub::protocol::common::thread_id::ThreadId as core::convert::TryFrom<&[u8]>>::try_from
0.1%   1.7%    254B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_specific_thread_id
0.0%   1.4%    212B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write
0.0%   0.9%    130B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::inner_write
0.0%   0.9%    128B           gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   0.8%    115B           gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   0.8%    111B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::flush
0.0%   0.7%    102B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.7%     97B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   0.7%     96B           gdbstub <gdbstub::protocol::common::thread_id::IdKind as core::convert::TryFrom<&[u8]>>::try_from
0.0%   0.6%     95B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   0.6%     88B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_hex
0.0%   0.5%     79B           gdbstub <usize as gdbstub::internal::be_bytes::BeBytes>::from_be_bytes
0.0%   0.5%     77B           gdbstub <u32 as gdbstub::internal::be_bytes::BeBytes>::from_be_bytes
0.0%   0.4%     65B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.4%     65B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.4%     65B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.4%     64B      gdbstub_arch <gdbstub_arch::arm::reg::arm_core::ArmCoreRegs as gdbstub::arch::Registers>::gdb_deserialize
0.0%   0.4%     57B compiler_builtins __rust_probestack
0.0%   0.3%     50B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadR...
0.0%   0.3%     50B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadR...
0.0%   0.3%     50B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadR...
0.0%   0.3%     38B         [Unknown] _start
0.0%   0.0%      6B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::breakpoints::SwBreakpoint>::add...
0.0%   0.0%      0B                   And 0 smaller methods. Use -n N to show more.
3.4% 100.0% 14.4KiB                   .text section size, the file size is 425.3KiB
target/release/gdbstub-nostd  :
section               size    addr
.interp                 28     792
.note.gnu.property      32     824
.note.gnu.build-id      36     856
.note.ABI-tag           32     892
.gnu.hash               36     928
.dynsym                360     968
.dynstr                204    1328
.gnu.version            30    1532
.gnu.version_r          64    1568
.rela.dyn              408    1632
.init                   27    4096
.plt                    16    4128
.plt.got                 8    4144
.text                14718    4160
.fini                   13   18880
.rodata                946   20480
.eh_frame_hdr          260   21428
.eh_frame             1316   21688
.init_array              8   28072
.fini_array              8   28080
.dynamic               448   28088
.got                   136   28536
.data                    8   28672
.bss                     8   28680
.comment                43       0
Total                19193
```

`cargo bloat` after:
```
    Analyzing target/release/gdbstub-nostd

File  .text    Size             Crate Name
2.5%  72.5% 10.6KiB         [Unknown] main
0.2%   4.8%    724B           gdbstub gdbstub::stub::state_machine::GdbStubStateMachineInner<gdbstub::stub::state_machine::stat...
0.1%   2.0%    301B           gdbstub gdbstub::protocol::commands::breakpoint::BasicBreakpoint::from_slice
0.1%   1.8%    274B           gdbstub gdbstub::stub::core_impl::resume::<impl gdbstub::stub::core_impl::GdbStubImpl<T,C>>::writ...
0.1%   1.8%    269B           gdbstub gdbstub::protocol::common::hex::decode_hex_buf
0.1%   1.7%    259B           gdbstub <gdbstub::protocol::common::thread_id::ThreadId as core::convert::TryFrom<&[u8]>>::try_from
0.1%   1.7%    254B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_specific_thread_id
0.0%   1.4%    212B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write
0.0%   0.9%    130B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::inner_write
0.0%   0.9%    128B           gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   0.8%    115B           gdbstub gdbstub::protocol::common::hex::decode_hex
0.0%   0.7%    111B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::flush
0.0%   0.7%    102B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.6%     97B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   0.6%     96B           gdbstub <gdbstub::protocol::common::thread_id::IdKind as core::convert::TryFrom<&[u8]>>::try_from
0.0%   0.6%     95B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_num
0.0%   0.6%     88B           gdbstub gdbstub::protocol::response_writer::ResponseWriter<C>::write_hex
0.0%   0.5%     79B           gdbstub <usize as gdbstub::internal::be_bytes::BeBytes>::from_be_bytes
0.0%   0.5%     77B           gdbstub <u32 as gdbstub::internal::be_bytes::BeBytes>::from_be_bytes
0.0%   0.4%     65B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.4%     65B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.4%     65B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadB...
0.0%   0.4%     64B      gdbstub_arch <gdbstub_arch::arm::reg::arm_core::ArmCoreRegs as gdbstub::arch::Registers>::gdb_deserialize
0.0%   0.4%     57B compiler_builtins __rust_probestack
0.0%   0.3%     50B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadR...
0.0%   0.3%     50B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadR...
0.0%   0.3%     50B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::base::multithread::MultiThreadR...
0.0%   0.3%     38B         [Unknown] _start
0.0%   0.0%      6B    gdbstub_nostd? <gdbstub_nostd::gdb::DummyTarget as gdbstub::target::ext::breakpoints::SwBreakpoint>::add...
0.0%   0.0%      0B                   And 0 smaller methods. Use -n N to show more.
3.4% 100.0% 14.6KiB                   .text section size, the file size is 432.8KiB
target/release/gdbstub-nostd  :
section               size    addr
.interp                 28     792
.note.gnu.property      32     824
.note.gnu.build-id      36     856
.note.ABI-tag           32     892
.gnu.hash               36     928
.dynsym                360     968
.dynstr                204    1328
.gnu.version            30    1532
.gnu.version_r          64    1568
.rela.dyn              408    1632
.init                   27    4096
.plt                    16    4128
.plt.got                 8    4144
.text                14977    4160
.fini                   13   19140
.rodata                954   20480
.eh_frame_hdr          260   21436
.eh_frame             1316   21696
.init_array              8   28072
.fini_array              8   28080
.dynamic               448   28088
.got                   136   28536
.data                    8   28672
.bss                     8   28680
.comment                43       0
Total                19460
```

xobs commented:
Note that we're able to save an additional 50 bytes by combining `Base::qC` with `Base::qfThreadInfo`, but I don't know if it's worth it in terms of complexity:

```diff
diff --git a/src/stub/core_impl/base.rs b/src/stub/core_impl/base.rs
index decc442..3ac9c3a 100644
--- a/src/stub/core_impl/base.rs
+++ b/src/stub/core_impl/base.rs
@@ -338,51 +338,13 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
                 }
                 HandlerStatus::NeedsOk
             }
-            Base::qC(_) => {
-                res.write_str("QC")?;
-                let pid = self.get_sane_current_pid(target)?;
-
-                match target.base_ops() {
-                    BaseOps::SingleThread(_) => res.write_specific_thread_id(SpecificThreadId {
-                        pid: self
-                            .features
-                            .multiprocess()
-                            .then_some(SpecificIdKind::WithId(pid)),
-                        tid: SpecificIdKind::WithId(SINGLE_THREAD_TID),
-                    })?,
-                    BaseOps::MultiThread(ops) => {
-                        let mut err: Result<_, Error<T::Error, C::Error>> = Ok(());
-                        let mut first = true;
-                        ops.list_active_threads(&mut |tid| {
-                            // TODO: replace this with a try block (once stabilized)
-                            let e = (|| {
-                                if !first {
-                                    return Ok(());
-                                }
-                                first = false;
-                                res.write_specific_thread_id(SpecificThreadId {
-                                    pid: self
-                                        .features
-                                        .multiprocess()
-                                        .then_some(SpecificIdKind::WithId(pid)),
-                                    tid: SpecificIdKind::WithId(tid),
-                                })?;
-                                Ok(())
-                            })();

-                            if let Err(e) = e {
-                                err = Err(e)
-                            }
-                        })
-                        .map_err(Error::TargetError)?;
-                        err?;
-                    }
+            Base::qfThreadInfo(_) | Base::qC(_) => {
+                if let Base::qfThreadInfo(_) = command {
+                    res.write_str("m")?;
+                } else if let Base::qC(_) = command {
+                    res.write_str("QC")?;
                 }
-
-                HandlerStatus::Handled
-            }
-            Base::qfThreadInfo(_) => {
-                res.write_str("m")?;
                 let pid = self.get_sane_current_pid(target)?;

                 match target.base_ops() {
@@ -400,6 +362,9 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
                             // TODO: replace this with a try block (once stabilized)
                             let e = (|| {
                                 if !first {
+                                    if let Base::qC(_) = command {
+                                        return Ok(());
+                                    }
                                     res.write_str(",")?
                                 }
                                 first = false;
```

xobs commented:
Were there other changes you want to see in this PR?

I'm stuck right now on https://github.com/renode/renode/issues/441https://github.com/renode/renode/issues/442 and https://github.com/renode/renode/issues/443 -- the emulator I use doesn't seem to handle short sequences between exceptions.

I'll see if I can't try it on real hardware soon.

daniel5151 reviewed (changes requested):
There's still that comment on the `vAttach` handler.

In the future, please refrain from resolving comments yourself. I'll resolve 'em myself once i've confirmed they LGTM. (I know there are different schools of thoughts here - this is just the camp that I fall into wrt. PR comments)

xobs commented:
Alright, as requested I moved the stop reporting into its own function, and am now using that. I also removed `S00` and found out hat the `T05thread` approach works:

```
(gdb) att 2
Attaching to program: E:\Code\Xous\Core\target\riscv32imac-unknown-xous-elf\release\xous-ticktimer, Remote target
[remote] Sending packet: $vAttach;2#38
[remote] Received Ack
[remote] Packet received: T05thread:02;
[remote] packet_ok: Packet vAttach (attach) is supported
[remote] Sending packet: $qC#b4
[remote] Received Ack
[remote] Packet received: QC02
[remote] Sending packet: $Hg2#e1
[remote] Received Ack
[remote] Packet received: OK
[remote] wait: enter
[remote] wait: exit
[remote] Sending packet: $g#67
[remote] Received Ack
[remote] Packet received: 00000000f4a3010030f6ff7f000000000000006048000000a252030030010020f0faff7f0100000048fdff7ff0faff7f10fbff7f7469636b040000000e058c3a000000000001002017000000333333330400000068040020b805002001000000feffffff040000000f0f0f0f01010101010000000500000000000000900000207e320300
[remote] Sending packet: $qOffsets#4b
[remote] Received Ack
[remote] Packet received:
[remote] Sending packet: $qSymbol::#5b
[remote] Received Ack
[remote] Packet received:
[remote] packet_ok: Packet qSymbol (symbol-lookup) is NOT supported
[remote] Sending packet: $m33262,2#cb
[remote] Received Ack
[remote] Packet received: 3571
[remote] Sending packet: $m33264,2#cd
[remote] Received Ack
[remote] Packet received: 06cf
[remote] Sending packet: $m33266,2#cf
[remote] Received Ack
[remote] Packet received: 22cd
[remote] Sending packet: $m33268,2#d1
[remote] Received Ack
[remote] Packet received: 26cb
[remote] Sending packet: $m3326a,2#fa
[remote] Received Ack
[remote] Packet received: 4ac9
[remote] Sending packet: $m3326c,2#fc
[remote] Received Ack
[remote] Packet received: 4ec7
[remote] Sending packet: $m3326e,2#fe
[remote] Received Ack
[remote] Packet received: 52c5
[remote] Sending packet: $m33270,2#ca
[remote] Received Ack
[remote] Packet received: 56c3
[remote] Sending packet: $m33272,2#cc
[remote] Received Ack
[remote] Packet received: 5ac1
[remote] Sending packet: $m33274,2#ce
[remote] Received Ack
[remote] Packet received: 2e84
[remote] Sending packet: $m33276,2#d0
[remote] Received Ack
[remote] Packet received: 0328
[remote] Sending packet: $m33278,2#d2
[remote] Received Ack
[remote] Packet received: 0600
[remote] Sending packet: $m3327a,2#fb
[remote] Received Ack
[remote] Packet received: 2320
[remote] Sending packet: $m3327c,2#fd
[remote] Received Ack
[remote] Packet received: 0600
[remote] Sending packet: $m3327a,2#fb
[remote] Received Ack
[remote] Packet received: 2320
[remote] Sending packet: $qfThreadInfo#bb
[remote] Received Ack
[remote] Packet received: m02,03
[remote] Sending packet: $qsThreadInfo#c8
[remote] Received Ack
[remote] Packet received: l
[New Thread 3]

Thread 1 stopped.
0x0003327e in xous::syscall::reply_and_receive_next_legacy ()
(gdb)
```

daniel5151 commented on src/stub/core_impl/base.rs:
```diff
@@ -34,6 +34,43 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+        }
+    }
+
+    // res.write_str("S00")?;
+    // HandlerStatus::Handled
```
this should go away

xobs replied:
Done

daniel5151 commented on src/stub/core_impl/base.rs:
```diff
@@ -34,6 +34,43 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+    // HandlerStatus::Handled
+
+    #[inline]
+    pub(crate) fn report_stop_reason(
```
```suggestion
    pub(crate) fn report_reasonable_stop_reason(
```

just so it's clear that it's not a "correct" stop reason - just a "reasonable" one

xobs replied:
Sure. Seems... reasonable.

daniel5151 commented on src/stub/core_impl/base.rs:
```diff
@@ -34,6 +34,43 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+    // res.write_str("S00")?;
+    // HandlerStatus::Handled
+
+    #[inline]
```
```suggestion
    #[inline(always)]
```

Empirically, I've found that bare `#[inline]` call aren't super useful - it's only with the addition of `always` or `never` that the compiler actually respects your requests. Otherwise, it's just a suggestion (that often gets ignored)

daniel5151 replied:
though, while you're here... can you double-check that my intuition here is correct?

i.e: is the code smaller if you force the inline, or does it actually make the binary larger?

xobs replied:
There's no difference between the two builds:
```sh
user@Ondo:/opt/Xous/gdbstub$ git diff
diff --git a/src/stub/core_impl/base.rs b/src/stub/core_impl/base.rs
index c392289..1cf42ee 100644
--- a/src/stub/core_impl/base.rs
+++ b/src/stub/core_impl/base.rs
@@ -44,7 +44,7 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
         }
     }

-    #[inline(always)]
+    #[inline]
     pub(crate) fn report_reasonable_stop_reason(
         &mut self,
         res: &mut ResponseWriter<'_, C>,
user@Ondo:/opt/Xous/gdbstub$ ./example_no_std/check_size.sh > inline.txt
   Compiling gdbstub v0.6.4 (/opt/Xous/gdbstub)
   Compiling gdbstub_arch v0.2.4 (/opt/Xous/gdbstub/gdbstub_arch)
   Compiling gdbstub-nostd v0.1.0 (/opt/Xous/gdbstub/example_no_std)
    Finished release [optimized + debuginfo] target(s) in 3.17s
    Finished release [optimized + debuginfo] target(s) in 0.00s
    Analyzing target/release/gdbstub-nostd

user@Ondo:/opt/Xous/gdbstub$ git checkout src/stub/core_impl/base.rs
Updated 1 path from the index
user@Ondo:/opt/Xous/gdbstub$ ./example_no_std/check_size.sh > inline-always.txt
   Compiling gdbstub v0.6.4 (/opt/Xous/gdbstub)
   Compiling gdbstub_arch v0.2.4 (/opt/Xous/gdbstub/gdbstub_arch)
   Compiling gdbstub-nostd v0.1.0 (/opt/Xous/gdbstub/example_no_std)
    Finished release [optimized + debuginfo] target(s) in 3.09s
    Finished release [optimized + debuginfo] target(s) in 0.00s
    Analyzing target/release/gdbstub-nostd

user@Ondo:/opt/Xous/gdbstub$ diff -u inline-always.txt inline.txt
user@Ondo:/opt/Xous/gdbstub$
```

daniel5151 commented on src/target/ext/extended_mode.rs:
```diff
@@ -95,6 +95,9 @@ pub trait ExtendedMode: Target {
     /// In all-stop mode, all threads in the attached process are stopped; in
     /// non-stop mode, it may be attached without being stopped (if that is
     /// supported by the target).
+    ///
+    /// After this method resolves, GDB will immediately request a stop reason.
+    /// It is up to the implementation to ensure the stop reason is accurate.
```
this doesn't apply anymore, right?

xobs replied:
I admit I wasn't entirely sure about that comment, but yes, that does not apply now so I have removed it.

daniel5151 commented on src/target/ext/base/multithread.rs:
```diff
@@ -110,6 +111,23 @@ pub trait MultiThreadBase: Target {
+    /// process.
+    ///
+    /// This should reflect the currently-debugged process which should be
+    /// updated when switching processes after calling `attach()`.
```
can you set up an intra-doc link that points at the specific `attach` method this is referring to? i.e: the one in the ExtendedMode IDET

xobs replied:
There is now an intra-doc link

daniel5151 reviewed (changes requested):
Broad strokes look good. Once these comments are resolved, I think we'll be good to go?

daniel5151 reviewed (changes requested):
Sorry to ask ya for another iteration, but it seems the `qC` implementation isn't quite right.

xobs commented on src/stub/core_impl/base.rs:
```diff
@@ -332,15 +349,59 @@ impl<T: Target, C: Connection> GdbStubImpl<T, C> {
+                            .then_some(SpecificIdKind::WithId(pid)),
+                        tid: SpecificIdKind::WithId(SINGLE_THREAD_TID),
+                    })?,
+                    BaseOps::MultiThread(ops) => {
```
What should it do if it is set to `SINGLE_THREAD_TID`? I've set it so that it does the previous behaviour of looking through all threads, but I can change it to something else if you prefer...

daniel5151 replied:
Nope, that's exactly the behavior I would've expected - you're good.

daniel5151 reviewed (approved):
Thanks for working with me through these iterations! This LGTM.

I'll try and find some time to cut a new 0.6.x with these changes in the next few days.

xobs commented:
Thanks for merging this! Would it be possible to create a new release?
